clang/test/SemaSYCL/union-kernel-param-neg.cpp

```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -fcxx-exceptions -verify -fsyntax-only %s
+
+// This test checks if compiler reports compilation error on an attempt to pass
+// accessor/sampler/pointer as SYCL kernel parameter inside union.
+
+#include "sycl.hpp"
+using namespace cl::sycl;
+
+union union_with_sampler {
+  cl::sycl::sampler smpl;
+  // expected-error@-1 {{'cl::sycl::sampler' cannot be used as the type of a kernel parameter}}
+};
+
+union union_with_pointer {
+  int *ptr_in_union;
+  // expected-error@-1 {{'int *' cannot be used as the type of a kernel parameter}}
+};
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void a_kernel(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+
+  using Accessor =
+      accessor<int, 1, access::mode::read_write, access::target::global_buffer>;
+
+  union union_with_accessor {
+    Accessor member_acc[1];
+    // expected-error@-1 {{'Accessor' (aka 'accessor<int, 1, access::mode::read_write, access::target::global_buffer>') cannot be used as the type of a kernel parameter}}
```

Comment:
Yes, the diagnostic message is not correct. 
```suggestion
    // expected-error@-1 {{'Accessor' (aka 'accessor<int, 1, access::mode::read_write, access::target::global_buffer>') cannot be used inside a union kernel parameter}}
```
It would be great to add stack trace that points to kernel invocation, like Erich did here https://github.com/intel/llvm/pull/2289

Reply:
We are going to the object stack-trace here in a subsequent patch.  We should do it for the field checker as well, so I think there should be 1 patch to do both.

Reply:
Yes, I will create a separate PR for this.